From c76485faaaf5969b2a60f4197c0a75a4fd7888bd Mon Sep 17 00:00:00 2001
From: Andy Ross <andrew.j.ross@intel.com>
Date: Wed, 22 Oct 2014 16:23:38 +0800
Subject: [PATCH 106/117] atomisp: Disable PCI runtime PM

On BYT, the ISP declares support for runtime PM in its PCI
configuration space, but no support exists, the field is simply wrong.
Use a fixup to disable the support or else the PCI layer will spam the
log with warnings about the device refusing state changes.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>
---
 arch/x86/platform/intel-mid/atomisp_gmin_platform.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/arch/x86/platform/intel-mid/atomisp_gmin_platform.c b/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
index cecdc096..d5aaf84 100644
--- a/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
+++ b/arch/x86/platform/intel-mid/atomisp_gmin_platform.c
@@ -2,6 +2,7 @@
 #include <linux/i2c.h>
 #include <linux/dmi.h>
 #include <linux/efi.h>
+#include <linux/pci.h>
 #include <linux/acpi.h>
 #include <linux/delay.h>
 #include <media/v4l2-subdev.h>
@@ -691,3 +692,16 @@ int camera_sensor_csi(struct v4l2_subdev *sd, u32 port,
         return 0;
 }
 EXPORT_SYMBOL_GPL(camera_sensor_csi);
+
+#ifdef CONFIG_GMIN_INTEL_MID
+/* PCI quirk: The BYT ISP advertises PCI runtime PM but it doesn't
+ * work.  Disable so the kernel framework doesn't hang the device
+ * trying.  The driver itself does direct calls to the PUNIT to manage
+ * ISP power. */
+static void isp_pm_cap_fixup(struct pci_dev *dev)
+{
+	dev_info(&dev->dev, "Disabling PCI power management on camera ISP\n");
+	dev->pm_cap = 0;
+}
+DECLARE_PCI_FIXUP_FINAL(PCI_VENDOR_ID_INTEL, 0x0f38, isp_pm_cap_fixup);
+#endif
-- 
1.9.3

